docs/source/caching/index.rst

```diff
@@ -0,0 +1,115 @@
+
+How are cached nodes matched?
+-----------------------------
+
+To determine wheter a given node is identical to an existing one, a hash of the content of the node is created. If a node of the same class with the same hash already exists in the database, this is considered a cache match. You can manually check the hash of a given node with the :meth:`.get_hash() <.AbstractNode.get_hash>` method.
+
+By default, this hash is created from:
+
+* all attributes of a node, except the ``_updatable_attributes``
+* the ``__version__`` of the module which defines the node class
+* the content of the repository folder of the node
+* the UUID of the computer, if the node has one
+
+In the case of calculations, the hashes of the inputs are also included. When developing calculation and data classes, there are some methods you can use to determine how the hash is created:
+
+* To ignore specific attributes, a ``Node`` subclass can have a ``_hash_ignored_attributes`` attribute. This is a list of attribute names which are ignored when creating the hash.
+* To add things which should be considered in the hash, you can override the :meth:`_get_objects_to_hash <.AbstractNode._get_objects_to_hash>` method. Note that doing so overrides the behavior described above, so you should make sure to use the ``super()`` method.
+* Pass a keyword argument to :meth:`.get_hash <.AbstractNode.get_hash>`. These are passed on to ``aiida.common.hashing.make_hash``. For example, the ``ignored_folder_content`` keyword is used by the :class:`JobCalculation <.AbstractJobCalculation>` to ignore the ``raw_input`` subfolder of its repository folder.
+
+Additionally, there are two methods you can use to disable caching for particular nodes:
+
+* The :meth:`._is_valid_cache` method determines whether a particular node can be used as a cache. This is used for example to disable caching from failed calculations.
+* Node classes have a ``_cacheable`` attribute, which can be set to ``False`` to completely switch off caching for nodes of that class. This avoids performing queries for the hash altogether.
+
+There are two ways in which the hash match can go wrong: False negatives, where two nodes should have the same hash but do not, or false positives, where two different nodes have the same hash. It is important to understand that false negatives are **highly preferrable**, because they only increase the runtime of your calculations, as if caching was disabled. False positives however can break the logic of your calculations. Be mindful of this when modifying the caching behaviour of your calculation and data classes.
+
+.. _caching_provenance:
+
```

Comment:
I would add something like the following section
```rst
What to do if caching is used but you don't want to
---------------------------------------------------
While this should not happen, there are a few edge cases where this could. One example is when one uses a parser that lives in a different python module than the class itself (maybe even set via `calc.set_parserclass`) or maybe one forgets to increase the version number of the python package (that is used inside the hash).

In general, the "correct" solution to this problem is to properly set a version number (and for calculation plugins that depend on external parsers, depend on an exact version of them, and increase your version number when you start depending on a newer version of the parser).

However, in development, this could happen. In this case, if you see a calculation that is not running as you would like (and using results from an older calculation, but you want it to run again):

1. Check if it has a `_aiida_cached_from` attribute. If it has not, then it is not a problem of caching.
2. Go to that calculation (using e.g. `cached_from_node = load_node(uuid)`), and call the method `cached_from_node.clear_hash()` (optional, as this is fixed by the next step).
3. Check if there are more nodes with the same hash: 
   `equivalent = Node.get_all_same_nodes(node.get_hash())`, and for each of them clear the hash::
  
  for node in equivalent:
      node.clear_hash()

4. run your calculation again: this time it should be run again

If you instead think it is a bug in the AiiDA implementation, open an issue (with enough information to be able to reproduce the error, otherwise it is hard for us to help you) in the AiiDA GitHub repository [link here].
```

Note: here I'm assuming that there is (otherwise we should add it) a `Node.clear_hash()`, probably just deleting the `hash` attribute.
I'm also assuming the existence of a `Node.get_all_same_nodes(hash)` (feel free to change the name) to get not any equivalent node, but a list of all nodes with a given hash (always a list, possibly empty - it's a simple query, but better to provide a method).

What do you think of this proposal? I think the usecase described at the beginning of the docs I wrote here explains a case where such a functionality would be useful.

Reply:
I added this section now, and added the two methods.

The ``get_all_same_nodes`` is implemented as an instance method, not a classmethod. This is because the query takes into account the Node class also, not only its hash.